### PR TITLE
Support WindowProxy return values in bindings

### LINKED ISSUE #10965: Support WindowProxy return values in bindings
Right now, we return the `Window` and grab its `WindowProxy` in bindings code; this is confusing. We should support returning `WindowProxy` (`BrowsingContext`) objects directly from DOM methods.

### PR DESCRIPTION
**Pull request by farodin91**
Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data:
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy --faster` does not report any errors
- [x] These changes fix #10965 (github issue number if applicable).

Either:
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11214)

### CODE REVIEW

**Comment by farodin91** [comment]
@jdm I have no idea to fix this, any tipp for me? Thanks!
```
windowproxy.rs:12:1: 1:1 error: cannot borrow immutable borrowed content as mutable
windowproxy.rs:12:1: 1:1 note: in this expansion of #[_generate_reflector] (windowproxy.rs)
``` (Complete output)

**Comment by jdm** [comment]
Rather than answer the question that was asked, I'm going to suggest that this isn't the correct way to solve this. The `BrowsingContext` type that [already exists](https://github.com/servo/servo/blob/3a28aad349473d7a9fcb6c5e4841abf3972fd733/components/script/dom/browsingcontext.rs#L31) is the type in Servo that corresponds to `WindowProxy`; we shouldn't have to create a separate interface and type for this work.

**Comment by farodin91** [comment]
I'll start over. I miss understood something.

**Comment by farodin91** [comment]
@jdm 
My idea:
1.  use this function https://github.com/servo/servo/blob/master/components/script/dom/browsingcontext.rs#L119.
2. uncomment this lines https://github.com/servo/servo/blob/master/components/script/dom/webidls/Window.webidl#L9-L10 & comment the following 
3. adding support of this WindowProxy in WebIDL like ArrayBufferView in this file https://github.com/servo/servo/blob/master/components/script/dom/bindings/codegen/parser/WebIDL.py
   * Possible as IDLWrapperType or IDLBuiltinType
4. adapt the CodeGen.py

Any tips or additions?

**Comment by jdm** [comment]
Instead of teaching the WebIDL parser about WindowProxy, I would prefer to try to teach our specific bindings about the WindowProxy -> BrowsingContext translation. The `window_proxy` method on BrowsingContext delegates to the object's reflector (`self.reflector().get_jsobject()`), so our code generation should be able to use to `BrowsingContext` values already since it understands objects with reflectors. We should try modifying `Bindings.conf` to teach it that a `WindowProxy` WebIDL type should correspond to a `BrowsingContext` Rust type.

**Comment by farodin91** [comment]
Currently, i'm looking to prevent generating Binding for the WebIDL type, like ChildNode.
After find a solution. I'm will teach that a `WindowProxy` WebIDL type should correspond to a `BrowsingContext` Rust type.

**Comment by asajeffrey** [comment]
Is this still active?

**Comment by farodin91** [comment]
Yes. During the last weeks i completed some tasks for my education.

**Comment by asajeffrey** [comment]
Cool!

**Comment by farodin91** [comment]
@jdm r?

**Comment by jdm** [comment]
I reviewed your changes and made some of my own. I'm going to make a pull request to your branch when I finish fixing the merge conflicts, since the HTMLIFrameElement changes took me by surprise.

**Comment by jdm** [comment]
Opened https://github.com/farodin91/servo/pull/1 with my changes.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 1]
```diff
@@ -5569,6 +5574,7 @@ def __init__(self, config, prefix, webIDLFile):
             'dom::bindings::str::{ByteString, DOMString, USVString}',
             'dom::bindings::trace::RootedVec',
             'dom::bindings::weakref::{DOM_WEAK_SLOT, WeakBox, WeakReferenceable}',
+            'dom::browsingcontext::BrowsingContext',
```
Comment (farodin91) [inline components/script/dom/bindings/codegen/CodegenRust.py]: I look for integrating this line in codegen. Not as hardcoded.

**Comment by jdm** [comment]
@farodin91 These changes look really good! I think we should go ahead and squash them into one and merge them.

**Comment by farodin91** [comment]
@jdm How to update tests?

**Comment by ConnorGBrewster** [comment]
@farodin91 https://github.com/servo/servo/tree/master/tests/wpt#updating-test-expectations

**Comment by farodin91** [comment]
@jdm 
It was impossible to update /html/dom/interfaces.html.
They completely crashed

**Comment by farodin91** [comment]
@jdm I made an failure during squashing.

File: components/script/build.rs [hunk 2]
```diff
@@ -10,4 +10,5 @@ fn main() {
         .status()
         .unwrap()
         .success());
+    //panic!();
```
Comment (KiChjang) [inline components/script/build.rs]: Please remove this.

**Comment by KiChjang** [comment]
You can manually update the test expectation file at `tests/wpt/metadata/html/dom/interfaces.html.ini`.

**Comment by KiChjang** [comment]
It's also possible you need a rebase to master in order to fix the unscopable failures.

**Comment by farodin91** [comment]
My problem is rebased to latest commit on servo/master and resolved merged conflicts.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 3]
```diff
@@ -5063,13 +5067,9 @@ def __init__(self, descriptor):
                 descriptor.shouldHaveGetConstructorObjectMethod()):
             cgThings.append(CGGetConstructorObjectMethod(descriptor))
 
-        unscopableNames = []
```
Comment (KiChjang) [inline components/script/dom/bindings/codegen/CodegenRust.py]: This line was removed.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 4]
```diff
@@ -5081,9 +5081,7 @@ def __init__(self, descriptor):
                     raise TypeError("Stringifier attributes not supported yet. "
                                     "See https://github.com/servo/servo/issues/7590\n"
                                     "%s" % m.location)
-                if m.getExtendedAttribute("Unscopable"):
-                    assert not m.isStatic()
-                    unscopableNames.append(m.identifier.name)
```
Comment (KiChjang) [inline components/script/dom/bindings/codegen/CodegenRust.py]: These all look like they're unintended deletions.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 5]
```diff
@@ -5163,15 +5165,7 @@ def __init__(self, descriptor):
...
 
-        haveUnscopables = False
         if not descriptor.interface.isCallback():
-            if unscopableNames:
-                haveUnscopables = True
-                cgThings.append(
-                    CGList([CGGeneric("const unscopable_names: &'static [&'static [u8]] = &["),
-                            CGIndenter(CGList([CGGeneric(str_to_const_array(name)) for
-                                               name in unscopableNames], ",\n")),
-                            CGGeneric("];\n")], "\n"))
```
Comment (KiChjang) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Same here.

File: components/script/dom/bindings/codegen/CodegenRust.py (hunk 4 above)
Comment (farodin91) [inline components/script/dom/bindings/codegen/CodegenRust.py]: I'll merge after it builds. ;)

**Comment by jdm** [comment]
@bors-servo: retry
* #11613 #11703

**Comment by farodin91** [comment]
https://github.com/servo/servo/commit/999e057cb754da1d456670996b13b337c2762c0f

**Comment by jdm** [comment]
@bors-servo: retry
* #11693

**Pull request merged**